pkg/controllers/jobset_controller.go

```diff
@@ -371,54 +405,39 @@ func (r *JobSetReconciler) resumeJobsIfNecessary(ctx context.Context, js *jobset
 		}
 		jobsFromRJob := replicatedJobToActiveJobs[replicatedJob.Name]
 		for _, job := range jobsFromRJob {
+			if !jobSuspended(job) {
+				return nil
+			}
 			if err := r.resumeJob(ctx, job, nodeAffinities); err != nil {
 				return err
 			}
+			numJobsResumed += 1
 		}
 		// If using in order startup policy, return early and wait for the replicated job to be ready.
-		if inOrderStartupPolicy(startupPolicy) {
-			// Add a condition to the JobSet indicating the in order startup policy is executing.
-			return r.ensureCondition(ctx, ensureConditionOpts{
-				jobset:           js,
-				eventType:        corev1.EventTypeNormal,
-				forceFalseUpdate: true,
-				condition:        inOrderStartupPolicyExecutingCondition(),
-			})
+		if numJobsResumed > 0 && inOrderStartupPolicy(startupPolicy) {
+			setInOrderStartupPolicyInProgress(js, updateStatusOpts)
+			return nil
 		}
 	}
+
+	// If no jobs were resumed / no action was taken, there's nothing more to do here.
+	if numJobsResumed == 0 {
+		return nil
+	}
```

Comment:
Why did we need this? should the set conditions below ensure that the condition is updated only if it really needed to anyway?

Reply:
Also, for conditions that sets the status to false, `updateCondition` should handle this by simply equating the nonexistence of the condition with one that requests to set it to false.

Reply:
>Why did we need this? should the set conditions below ensure that the condition is updated only if it really needed to anyway?

The main reconcile loop doesn't distinguish between a jobset that has never been suspended and a jobset that was previously suspended and now needs to be resumed, so it always enters the `resumeJobSetIfNecessary()` function. This resulted in the "JobSet resumed" condition being set on jobsets that were never suspended, so I checked here if we actually had suspended jobs that were resumed before adding that condition.

Reply:
but then we can fix that via the below suggestion, which should fix it for all cases, not just here.

> Also, for conditions that sets the status to false, `updateCondition` should handle this by simply equating the nonexistence of the condition with one that requests to set it to false.

Reply:
> Also, for conditions that sets the status to false, updateCondition should handle this by simply equating the nonexistence of the condition with one that requests to set it to false.

If I understand your suggestion correctly, we could do this but we would have to change how we set the "in-order startup policy in progress" condition.  

We currently add a new condition with status of false to track that an in-order startup policy is currently in progress but has not completed yet. Once it completed, we update that false status to true. So if we want to maintain this current logic, we can't treat the condition not existing and a false condition as equivalent, otherwise the initial "in progres" (status = false) condition wouldn't be set.

I recently suggested to @kannon92 we migrate to using 2 separate conditions for tracking "in progress" vs "completed" in-order startup policy, since I personally find the false->true thing a bit unintuitive. What do you think?

Reply:
Yes, I think we should have two: InProgress and Complete and they should be mutually exclusive. See how we did something similar in LWS: https://github.com/kubernetes-sigs/lws/blob/45854111b905cc2acc474cda34dd9e9fc6a293de/pkg/controllers/leaderworkerset_controller.go#L557

Reply:
Sounds good, I refactored to split in order startup policy condition into 2 separate conditions, "in progress" and "copmleted." This is ready for another look.